### PR TITLE
Improve Helper class with adding tests

### PR DESCRIPTION
**Pull request by jinseoung-lee**
Hi
i thought i could improve the helper class by looking  at the code for about building queries.

All changes for lazy processing with linq
- `AllIndexesOf`, use `yield` statements
- `ReplaceAll`, use LINQ `Aggregate` method
- `Flatten`, use LINQ `SelectMany`

and do you have any test naming rule? maybe if you don't like this style, just comment me 😄

### CODE REVIEW

File: QueryBuilder/Helper.cs [hunk 1]
```diff
@@ -83,57 +62,40 @@ public static bool IsSubclassOfRawGeneric(Type generic, Type toCheck)
                 var cur = IsGenericType(toCheck) ? toCheck.GetGenericTypeDefinition() : toCheck;
 
                 if (generic == cur)
-                {
```
Comment (ahmad-moussawi) [inline QueryBuilder/Helper.cs]: one-line braces

File: QueryBuilder/Helper.cs [hunk 2]
```diff
@@ -83,57 +62,40 @@ public static bool IsSubclassOfRawGeneric(Type generic, Type toCheck)
...
-
-            for (var i = 1; i < tokens.Count; i++)
-            {
-                var replacement = callback.Invoke(i - 1);
-                newStr.Add(replacement + tokens[i]);
-            }
 
-            return string.Join("", newStr);
+            var index = 0;
+            return subject
```
Comment (ahmad-moussawi) [inline QueryBuilder/Helper.cs]: feeling that statements like this are harder to understand 🙄

File: QueryBuilder/Helper.cs [hunk 3]
```diff
@@ -25,36 +25,15 @@ public static bool IsNumber(object value)
 
         public static bool IsArray(object value)
         {
-            if (value == null)
-            {
-                return false;
-            }
-
             if (value is string)
-            {
```
Comment (ahmad-moussawi) [inline QueryBuilder/Helper.cs]: leaving the braces is more preferable https://softwareengineering.stackexchange.com/questions/2715/should-curly-braces-appear-on-their-own-line

Comment (jinseoung-lee) [reply QueryBuilder/Helper.cs]: ok, i agree 😉, what your choice both below options?

```csharp
// A
if (statement) { /* do something */ }

// B
if (statement)
{
    /* do something */
}
```

In project, several guard clause are exists without braces. i thought i could get rid of them. 😅

File: QueryBuilder/Helper.cs (hunk 2 above)
Comment (jinseoung-lee) [reply QueryBuilder/Helper.cs]: Ok, I will refactor with `Select`, `StringBuilder`

File: QueryBuilder/Helper.cs (hunk 3 above)
Comment (ahmad-moussawi) [reply QueryBuilder/Helper.cs]: I think B is more appropriate, maybe we have to use a tool such StyleCop or something similar :)

**Pull request merged**